sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/util/common/CounterNameAndMetadata.java

```diff
@@ -0,0 +1,155 @@
+   */
+  public static CounterNameAndMetadata msecs(String name) {
+    return named(name + "-msecs");
+  }
+
+  /**
+   * Returns a CounterName identical to this, but with the given origin.
+   */
+  public CounterNameAndMetadata withOrigin(String origin) {
+    return new CounterNameAndMetadata(this.name, origin, this.stepName, this.prefix);
+  }
+
+  /**
+   * Returns a CounterName identical to this, but with the given original step name.
+   */
+  public CounterNameAndMetadata withOriginalStepName(String originalStepName) {
+    return new CounterNameAndMetadata(this.name, this.origin, originalStepName, this.prefix);
+  }
+
+  /**
+   * Returns a CounterName identical to this, but with the given prefix.
+   */
+  public CounterNameAndMetadata withPrefix(String prefix) {
+    return new CounterNameAndMetadata(this.name, this.origin, this.stepName, prefix);
+  }
+
+  /**
+   * Name of the counter.
+   *
+   * <p>For example, process-msecs, ElementCount.
+   */
+  private final String name;
+
+  /**
+   * Origin (namespace) of counter name.
+   *
+   * <p>For example, "user" for user-defined counters.
+   * It may be null or empty for counters defined by the Dataflow service or SDK.
+   */
+  @Nullable
+  private final String origin;
+
+  /**
+   * System defined step name.
+   *
+   * <p>For example, s1, s2.out.
+   * It may be null or empty when counters don't associate with step names.
+   */
+  @Nullable
+  private final String stepName;
+
+  /**
+   * Prefix of group of counters.
+   *
+   * <p>It is null or empty when counters don't have general prefixes.
+   */
+  @Nullable
+  private final String prefix;
+
+  /**
+   * Flat name is the equivalent unstructured name.
+   *
+   * <p>It is null before {@link #getFlatName()} is called.
+   */
+  @Nullable
+  private String flatName;
+
+  private CounterNameAndMetadata(String name, String origin, String stepName, String prefix) {
+    this.name = name;
+    this.origin = origin;
+    this.stepName = stepName;
+    this.prefix = prefix;
+    this.flatName = null;
+  }
+
+  /**
+   * Returns the flat name of a structured counter.
+   */
+  public synchronized String getFlatName() {
+    if (flatName == null) {
+      StringBuilder sb = new StringBuilder();
+      if (!Strings.isNullOrEmpty(prefix)) {
+        // Flume doesn't explicitly use "-" to concatenate prefix, it may already have it in it.
+        sb.append(prefix);
+      }
+      if (!Strings.isNullOrEmpty(origin)) {
+        sb.append(origin + "-");
+      }
+      if (!Strings.isNullOrEmpty(stepName)) {
+        sb.append(stepName + "-");
+      }
+      sb.append(name);
+      flatName = sb.toString();
+    }
+    return flatName;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    } else if (o instanceof CounterNameAndMetadata) {
+      CounterNameAndMetadata that = (CounterNameAndMetadata) o;
+      return this.getFlatName().equals(that.getFlatName());
```

Comment:
This means that calling equals will actually have a side-effect (of constructing the flat name). Why not write it as an equality on the individual pieces of the CounterName?